### PR TITLE
Reserve Order ID for the child quote, copy it to parent just before the quote to order submit

### PR DESCRIPTION
**Pull request by danac-gs**
Some 3rd party plugins read Quote::getReservedOrderId() value and if set consider the quote to be a complete order. It is set by calling Quote::reserveOrderId(). We used to call it on the parent quote before creating the immutable quote. Now it is called on the first created immutable quote and stored in the separate field added to the quote model to be set on the parent quote just before order creation.

### CODE REVIEW

File: Setup/UpgradeSchema.php [hunk 1]
```diff
@@ -46,6 +46,17 @@ public function upgrade(SchemaSetupInterface $setup, ModuleContextInterface $con
             ]
         );
 
+        $setup->getConnection()->addColumn(
```
Comment (daisy1754) [inline Setup/UpgradeSchema.php]: is reservedOrderId field indexed? if so maybe we should index this too

File: Helper/Cart.php [hunk 2]
```diff
@@ -526,6 +526,56 @@ public function replicateQuoteData($source, $destination)
...
+    protected function reserveOrderId($immutableQuote, $quote)
+    {
+        $reservedOrderId = $immutableQuote->getBoltReservedOrderId();
+        if (!$reservedOrderId) {
+            $reservedOrderId = $immutableQuote->reserveOrderId()->getReservedOrderId();
+            $immutableQuote->setBoltReservedOrderId($reservedOrderId);
+            $quote->setBoltReservedOrderId($reservedOrderId);
+            $this->quoteResourceSave($quote);
+        } else {
+            $immutableQuote->setReservedOrderId($reservedOrderId);
```
Comment (daisy1754) [inline Helper/Cart.php]: I'm not sure if I follow here. what is this else block for?

File: Setup/UpgradeSchema.php (hunk 1 above)
Comment (danac-gs) [reply Setup/UpgradeSchema.php]: Index is not required we just do set and get on a loaded model, no searches or queries of any kind.

File: Helper/Cart.php (hunk 2 above)
Comment (danac-gs) [reply Helper/Cart.php]: reserveOrderId() is called on the first immutable quote and then propagated to parent and to subsequent immutables in BoltReservedOrderId field and in the end copied from parent BoltReservedOrderId to ReservedOrderId. This else is for consistency, to keep every subsequent immutable having all the fields populated. Actually, one more important reason, some lines below $cart['display_id'] = $immutableQuote->getReservedOrderId() . ' / ' . $immutableQuote->getId(); It is expected to be there. So every call to getReservedOrderId remain unchanged.

**Pull request merged**